[EventHubs] Add EventProcessor<TPartition> Sample from Upcoming Blog Post (#18434)
Azure/azure-sdk-for-net: sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample08_EventProcessor.md

# Building a custom processor with EventProcessor

This sample demonstrates using `EventProcessor<TPartition>` to build a custom event processor which manages its own load balancing and checkpoint state and then shows how to use it to consume events from Event Hubs. For most users we strongly recommend using the [EventProcessorClient](https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples) from the [Azure.Messaging.EventHubs.Processor](https://www.nuget.org/packages/Azure.Messaging.EventHubs.Processor) package instead of implementing your own processor library.

## What does a processor do?

> When consuming events, the consumer is tied to a specific event hub partition and reads the events in order from that partition. This means, to read all events from an Event Hub, you'll want to have one consumer per partition reading and processing events. For some workloads (where the processing can be intensive) you may want these consumers spread out across multiple machines, or even add or remove machines as the rate of incoming events increases or decreases. We call this process *load balancing*. In addition, you'll want to make sure that if one of your consumers has a problem (say the machine it is running on crashes) you're able to resume your processing at a later time, picking up where you left off. We call this process *checkpointing*.

Reviewer: ```suggestion
When consuming events, the consumer is tied to a specific partition of an Event Hub and reads the events in order from that partition. This means, to read all events from an Event Hub, you'll need one consumer per partition reading and processing events. For some workloads (where the processing can be intensive) you may want these consumers spread out across multiple machines, or even to add and remove consumers as the rate of incoming events increases or decreases. We call this process *load balancing*. In addition, you'll want to make sure that if one of your consumers has a problem (such as the machine it is running on crashes) you're able to resume your processing at a later time, picking up where you left off. We call this process *checkpointing*.
```

You may want to consider linking the `load balancing` definition to [this](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/Sample04_ProcessingEvents.md#load-balancing) write up for more context.   Similarly, there's a blurb on `checkpointing` found [here](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/Sample04_ProcessingEvents.md#checkpointing-while-processing) that may help with additional context.

> While you could write code directly using the `EventHubConsumerClient` to handle these cases, the Azure SDK also provides a type, `EventProcessorClient`, which handles this for you. It uses the Azure Blob Storage SDK to handle storing the state that's used during the load balancing and checkpointing operations. For many customers, that's sufficient, but some customers have asked for the ability to control how the information related to checkpointing and load balancing is stored. But for customers who have unique needs, we've exposed the lower-level machinery needed to build your own processor, which gives you full control over how this data is stored.

Reviewer: You may want to mention the `PartitionReceiver` here as well - or maybe not?

## Building an Event Processor with `EventProcessor<TPartition>`

The low level machinery needed to build you own processor, with full control over how its state is stored, is exposed by the `EventProcessor<TPartition>` class in the `Azure.Messaging.EventHubs.Primitives` namespace. This is an abstract class, which has five methods you'll need to provide implementations for. If you've used the `EventProcessorClient`, the first two will be familiar to you:

> - `OnProcessingEventBatchAsync`: The actual "business logic" of your processor. This is similar to the `ProcessEvent` event exposed by `EventProcessorClient`.

Reviewer: I'd suggest calling out the `OnInitializingPartition` and `OnPartitionProcessingStopped` methods as well.    Based on the conversations that I've had while helping teams implement custom processors, I'd assume they'll be used by most of the advanced scenarios for health monitoring and ensuring a checkpoint when a partition is relinquished.

- `OnProcessingErrorAsync`: A handler that can be used to observe exceptions from inside the machinery of the EventProcessor itself. This is similar to the `ProcessError` event exposed by `EventProcessorClient`.

The remaining three deal with checkpointing and load balancing.

- `ListOwnershipAsync`: Used by load balancing to see which partitions have been assigned to which processors.
- `ClaimOwnershipAsync`: Used by load balancing to grab ownership of a partition (which may or may not be assigned to another processor).

> - `ListCheckpointsAsync`: Used to find the starting point inside a partition when a processor starts reading events from a partition.

Reviewer: I'd strongly suggest talking about [`GetCheckpointAsync`](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor.cs#L806) and recommending that folks no-op `ListCheckpointsAsync` where possible.    That will allow them to return a specific checkpoint for the partition rather than the less efficient search through the collection on each iteration.

I have no idea why we originally implemented it as a "look through the list" but never cached it, but we recently added the optimization to help with performance.

The two `On*` methods can only be implemented when we know what we want our event processor to do, but the remaining can be implemented independent of the business logic for processing events.

In this sample, we will use the same storage mechanism as EventProcessorClient, i.e. we will store checkpoint and load balancing state using Azure Blob Storage.

> ```C# Snippet:EventHubs_Sample08_AzureBlobStorageEventProcessor

Reviewer: We probably want to give the line length a look here as well.   We've gotten feedback from customers letting us know that they like to read our samples directly on the GitHub site and asking that we be careful that the code lines aren't too long for that notoriously small area.

                OwnerIdentifier = blob.Metadata[OwnerIdentifierMetadataKey],
                PartitionId = blob.Name.Substring(ownershipBlobsPefix.Length),
                Version = blob.Properties.ETag.ToString()
            });
        }

        return partitonOwnerships;
    }

    // To claim ownership of a partition, we have to write a blob to blob storage (if this partition has never been claimed before) or update the metadata of an existing blob.
    protected override async Task<IEnumerable<EventProcessorPartitionOwnership>> ClaimOwnershipAsync(IEnumerable<EventProcessorPartitionOwnership> desiredOwnership, CancellationToken cancellationToken = default)
    {
        List<EventProcessorPartitionOwnership> claimedOwnerships = new List<EventProcessorPartitionOwnership>();
    
        foreach (EventProcessorPartitionOwnership ownership in desiredOwnership)
        {
            Dictionary<string, string> ownershipMetadata = new Dictionary<string, string>()
            {
                { OwnerIdentifierMetadataKey, ownership.OwnerIdentifier },
            };

            // Construct the path to the blob and get a blob client for it so we can interact with it.

>             string ownershipBlob = string.Format(OwnershipPrefixFormat + ownership.PartitionId, ownership.FullyQualifiedNamespace.ToLowerInvariant(), ownership.EventHubName.ToLowerInvariant(), ownership.ConsumerGroup.ToLowerInvariant());

Reviewer: nit:  seems odd to have a string concatenation embedded in a `string.Format` call.

                    {
                        ConsumerGroup = ownership.ConsumerGroup,
                        EventHubName = ownership.EventHubName,
                        FullyQualifiedNamespace = ownership.FullyQualifiedNamespace,
                        LastModifiedTime = info.LastModified,
                        OwnerIdentifier = ownership.OwnerIdentifier,
                        PartitionId = ownership.PartitionId,
                        Version = info.ETag.ToString()
                    });
                }
            }
            catch (RequestFailedException e) when (e.ErrorCode == BlobErrorCode.BlobAlreadyExists || e.ErrorCode == BlobErrorCode.ConditionNotMet)
            {
                // In this case, another host has claimed the partition before we did. That's safe to ignore. We'll still try to claim other partitions.
            }
        }

        return claimedOwnerships;
    }

    private const string CheckpointPrefixFormat = "{0}/{1}/{2}/checkpoint/";
    private const string OffsetMetadataKey = "offset";

    // We use the same strategy for recording checkpoint information as ownership information (metadata on a blob in blob storage)

>     protected override async Task<IEnumerable<EventProcessorCheckpoint>> ListCheckpointsAsync(CancellationToken cancellationToken = default)

Reviewer: Storage Reference:  [GetCheckpointAsync](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/BlobCheckpointStore/BlobsCheckpointStore.cs#L363)
Processor Reference: [GetCheckpointAsync](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs#L797)

        {
            { OffsetMetadataKey, data.Offset.ToString(CultureInfo.InvariantCulture) },
        };

        using MemoryStream emptyStream = new MemoryStream(Array.Empty<byte>());
        await StorageContainer.GetBlobClient(checkpointBlob).UploadAsync(emptyStream, metadata: checkpointMetadata, cancellationToken: cancellationToken).ConfigureAwait(false);
    }        
}
```

## Using the Custom Processor

The `AzureBlobStorageEventProcessor<TPartiton>` above is abstract because we did not implement any of the `On*` methods. By separating the logic for state management and event processing, we can leverage `AzureBlobStorageEventProcessor<TPartiton>` to build many different processors, each with their own business logic, by extending `AzureBlobStorageEventProcessor<TPartiton>` and implementing `OnProcessingEventBatchAsync` and `OnProcessingErrorAsync`.

Here's a simple processor that just counts the number of events received in a batch and then creates a checkpoint so they are not processed again:

```C# Snippet:EventHubs_Sample08_CustomProcessor
class CustomProcessor : AzureBlobStorageEventProcessor<EventProcessorPartition>
{

>     public CustomProcessor(int eventBatchMaximumCount, string consumerGroup, string fullyQualifiedNamespace, string eventHubName, TokenCredential credential, BlobContainerClient storageContainer, EventProcessorOptions options = null) : base(eventBatchMaximumCount, consumerGroup, fullyQualifiedNamespace, eventHubName, credential, storageContainer, options)

Reviewer: You may want to expand on this by caching the last event processed in a custom partition and using that to emit a checkpoint on the `OnPartitionProcessingStopped` method.

    {
    }

>     protected async override Task OnProcessingEventBatchAsync(IEnumerable<EventData> events, EventProcessorPartition partition, CancellationToken cancellationToken)

Reviewer: I think it's important to discuss the "you need to handle any errors in your code" from these methods and demonstrate with a `try/catch` that wraps the entire method for all the OnXXX handlers.     We have some [common guidance](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/Sample03_EventProcessorHandlers.md#exceptions-in-handlers) in the processor samples that may be helpful.

>     protected async override Task OnProcessingEventBatchAsync(IEnumerable<EventData> events, EventProcessorPartition partition, CancellationToken cancellationToken)

Reviewer: You may want to mention that this method will be invoked concurrently so the implementation needs to be thread-safe and that there will only be one active invocation for any partition to ensure that the events from the partition arrive in order.

Reviewer: Maybe also that there will be one instance of `TPartition` created per partition and managed by the processor.  That instance will be passed each time for the partition, so the handler methods can count on a consistent instance.   If you're trying to preserve application state, the `TPartition` instance is a good place to do so.

    {
        Console.WriteLine($"Received batch of {events.Count()} events for partition {partition.PartitionId}");

>         if (events.Any())

Reviewer: I'd  urge caution here;  we've seen a lot of folks checkpoint more aggressively than needed.   We should probably consider a comment here calling out the trade-off for checkpointing and mentioning that it's not usually a good idea to do it on every event/batch.   Most advanced customers seem to use a blended approach of after XX events or YY amount of time has elapsed.

Reviewer: nit:  You're also demonstrating potentially enumerating the set multiple times, which we know is safe in this instance but is usually discouraged for `IEnumerable`.   Not sure that its worth changing, but wanted to mention it for consideration.

        {
            await CheckpointAsync(partition, events.Last(), cancellationToken);
        }
    }

    protected async override Task OnProcessingErrorAsync(Exception exception, EventProcessorPartition partition, string operationDescription, CancellationToken cancellationToken)
    {
        if (partition != null)
        {
            Console.Error.WriteLine($"Exception on partition {partition.PartitionId} while performing {operationDescription}: {exception.Message}");
        }
        else
        {
            Console.Error.WriteLine($"Exception while performing {operationDescription}: {exception.Message}");
        }        
    }    
}
```

This should look familiar if you've used `EventProcessorClient` before, but there are a few small differences you need to be aware of:

> - The processor works in batches, so when you construct it you need to specify the maximum number of events you'll want to process at once, using the `eventBatchMaximumCount` parameter. The `IEnumerable<EventData>` that is passed `OnProcessingEventBatchAsync` to will never contain more items than this maximum count.

Reviewer: But it may often contain less.   It is not unusual (though it is confusing) to see uneven sizes despite having a good number of events in the partition; this is due to how the AMQP library's prefetch cache works and is sensitive to network speed and other resource use.    That tends to confuse the heck out of people.

> - If there are fewer unprocessed messages than `eventBatchMaximumCount` the processor also contains a maximum amount of time it will wait for more events to show up. By default, this is 60 seconds, but can be controlled by setting the `MaximumWaitTime` property of the `EventProcessorOptions` parameter. If this is set to `null` it means that the processor will until it has received `eventBatchMaximumCount` events, otherwise it will deliver a batch that has fewer events in it.

Reviewer: You may want to add spacing between the bullets.   I found the condensed text a bit difficult to read.

> - If `MaximumWaitTime` is set, `OnProcessingEventBatchAsync` will be called even if no events have been consumed. In this case, the `IEnumerable<EventData>` will contain zero events (so we have to guard our call to `.Last()` which fails if there is no event.)

Reviewer: It may be helpful to mention that this is often useful for `heartbeat` semantics and creating a checkpoint on a time interval.
